File: harvester/audit.ts
import { appendFileSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { scanRepo } from './collect.ts';
import { probeAll, type Signal } from './probe.ts';
import { classify, fold, loadBaseline, saveBaseline, type Baseline, type Verdict } from './baseline.ts';

/**
 * The harvester entrypoint.
 *
 *   node harvester/audit.ts                # scan repo + probe every cited source
 *   node harvester/audit.ts --no-network   # repo scan only (deterministic, offline)
 *   node harvester/audit.ts --strict       # exit 1 if any source CHANGED (for gating)
 *
 * It scans the committed data tree, probes every official source URL the data
 * cites, and diffs each against harvester/baseline.json to flag CHANGED /
 * NEW / UNREACHABLE. It writes harvester/report.md and, in GitHub Actions,
 * the job summary. By design it exits 0 even when sources are unreachable or
 * have changed — those are findings to review, not job failures. Only
 * --strict turns a detected change into a non-zero exit.
 */

const ARGS = new Set(process.argv.slice(2));
const NO_NETWORK = ARGS.has('--no-network');
const STRICT = ARGS.has('--strict');
const ROOT = process.cwd();
const DATA_DIR = process.env.HARVEST_DATA_DIR ?? join(ROOT, 'data');
const BASELINE_PATH = process.env.HARVEST_BASELINE ?? join(ROOT, 'harvester', 'baseline.json');
const REPORT_PATH = process.env.HARVEST_REPORT ?? join(ROOT, 'harvester', 'report.md');
const STALE_DAYS = Number(process.env.HARVEST_STALE_DAYS ?? 400);

const ICON: Record<Verdict, string> = { new: '🆕', unchanged: '✅', changed: '⚠️', unreachable: '🔌' };
const clip = (s: string | undefined, n = 90): string => (s && s.length > n ? s.slice(0, n - 1) + '…' : s ?? '');

function daysSince(iso?: string): number | undefined {
  if (!iso) return undefined;
  const t = Date.parse(iso);
  return Number.isNaN(t) ? undefined : Math.floor((Date.now() - t) / 86400000);
}

interface Row { verdict: Verdict; url: string; signal: Signal; citedBy: string[]; title?: string; }

function buildReport(rows: Row[], filesScanned: number, parseErrors: string[], stale: string[]): string {
  const n = (v: Verdict) => rows.filter((r) => r.verdict === v).length;
  const changed = rows.filter((r) => r.verdict === 'changed');
  const unreachable = rows.filter((r) => r.verdict === 'unreachable');
  const now = new Date().toISOString().slice(0, 16).replace('T', ' ');
  const L: string[] = [];
  L.push('# 🌾 Omnia data harvester', '', `_Run ${now} UTC · ${filesScanned} data files · ${rows.length} cited sources_`, '');
  L.push('| | count |', '|---|--:|',
    `| ⚠️ changed | ${n('changed')} |`,
    `| 🆕 new | ${n('new')} |`,
    `| ✅ unchanged | ${n('unchanged')} |`,
    `| 🔌 unreachable | ${n('unreachable')} |`,
    `| 📄 JSON parse errors | ${parseErrors.length} |`,
    `| 🕰️ stale (>${STALE_DAYS}d) | ${stale.length} |`, '');

  if (changed.length) {
    L.push('## ⚠️ Sources that CHANGED since last run', '', 'A change here means the official page/PDF differs from what we last saw — review whether our data needs updating.', '');
    for (const r of changed) {
      L.push(`- **${clip(r.title) || r.url}**`, `  - ${r.url}`, `  - cited by: ${r.citedBy.join(', ')}`);
    }
    L.push('');
  }
  if (parseErrors.length) {
    L.push('## 📄 JSON parse errors (fix these — the data file is broken)', '', ...parseErrors.map((p) => `- ${p}`), '');
  }
  if (unreachable.length) {
    L.push('<details><summary>🔌 Unreachable this run (not a failure — retried next run)</summary>', '');
    for (const r of unreachable) L.push(`- ${clip(r.title) || r.url} — ${r.signal.error ?? 'unreachable'}`);
    L.push('', '</details>', '');
  }
  if (stale.length) {
    L.push('<details><summary>🕰️ Data files whose freshness marker is older than ' + STALE_DAYS + ' days</summary>', '', ...stale.map((s) => `- ${s}`), '', '</details>', '');
  }
  L.push('---', '_The harvester never fails on unreachable sources or detected changes; both are findings above. It only reports._');
  return L.join('\n');
}

async function main(): Promise<void> {
  const scan = scanRepo(DATA_DIR);
  const parseErrors = scan.files.filter((f) => !f.ok).map((f) => `${f.path}: ${f.parseError}`);
  const stale = scan.files
    .filter((f) => f.ok && f.asOf && (daysSince(f.asOf) ?? 0) > STALE_DAYS)
    .map((f) => `${f.path} (asOf ${f.asOf}, ${daysSince(f.asOf)}d ago)`);

  const urls = scan.sources.map((s) => s.url);
  const signals: Signal[] = NO_NETWORK
    ? urls.map((url) => ({ url, ok: false, error: 'network skipped (--no-network)' }))
    : await probeAll(urls, { concurrency: 10, timeoutMs: 15000, retries: 1 });

  const baseline: Baseline = loadBaseline(BASELINE_PATH);
  const now = new Date().toISOString();
  const rows: Row[] = [];
  for (let i = 0; i < scan.sources.length; i++) {
    const src = scan.sources[i];
    const sig = signals[i];
    const verdict = NO_NETWORK ? 'unreachable' : classify(baseline[src.url], sig);
    rows.push({ verdict, url: src.url, signal: sig, citedBy: src.citedBy, title: src.title });
    // Only evolve the baseline on a real network run, so --no-network never rewrites it.
    if (!NO_NETWORK) baseline[src.url] = fold(baseline[src.url], sig, now, verdict);
  }

  const report = buildReport(rows, scan.files.length, parseErrors, stale);
  writeFileSync(REPORT_PATH, report + '\n');
  if (process.env.GITHUB_STEP_SUMMARY) {
    try { appendFileSync(process.env.GITHUB_STEP_SUMMARY, report + '\n'); } catch { /* summary is best-effort */ }
  }
  if (!NO_NETWORK) saveBaseline(BASELINE_PATH, baseline);

  const count = (v: Verdict) => rows.filter((r) => r.verdict === v).length;
  console.log(`\n🌾 Harvester: ${scan.files.length} files, ${rows.length} sources — ⚠️ ${count('changed')} changed · 🆕 ${count('new')} new · ✅ ${count('unchanged')} unchanged · 🔌 ${count('unreachable')} unreachable · 📄 ${parseErrors.length} parse errors · 🕰️ ${stale.length} stale`);
  console.log(`   report → ${REPORT_PATH}`);

  if (STRICT && count('changed') > 0) {
    console.error('   --strict: sources changed; exiting non-zero.');
    process.exitCode = 1;
  }
}

main().catch((e) => {
  // Even an unexpected internal error is reported without killing the workflow,
  // unless --strict is set. "No fail" is the contract.
  console.error('harvester: unexpected error:', (e as Error)?.message ?? e);
  if (STRICT) process.exitCode = 1;
});

File: harvester/audit.py
#!/usr/bin/env python3
"""The harvester entrypoint.

    python harvester/audit.py                # scan repo + probe every cited source
    python harvester/audit.py --no-network   # repo scan only (deterministic, offline)
    python harvester/audit.py --strict       # exit 1 if any source is CONFIRMED changed

Scans the committed data tree, probes every official source URL it cites, and
uses a two-run debounce to report only CONFIRMED changes (content that appeared
on two consecutive runs). Writes harvester/report.md and, in GitHub Actions, the
job summary. Exits 0 even when sources are unreachable or changed — those are
findings, not failures. Only --strict turns a confirmed change into exit 1.
"""
from __future__ import annotations

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from collect import scan_repo          # noqa: E402
from probe import probe_all            # noqa: E402
from baseline import classify, fold, load_baseline, save_baseline, now_iso  # noqa: E402

ROOT = os.getcwd()
ARGS = set(sys.argv[1:])
NO_NETWORK = "--no-network" in ARGS
STRICT = "--strict" in ARGS
DATA_DIR = os.environ.get("HARVEST_DATA_DIR", os.path.join(ROOT, "data"))
BASELINE_PATH = os.environ.get("HARVEST_BASELINE", os.path.join(ROOT, "harvester", "baseline.json"))
REPORT_PATH = os.environ.get("HARVEST_REPORT", os.path.join(ROOT, "harvester", "report.md"))
STALE_DAYS = int(os.environ.get("HARVEST_STALE_DAYS", "400"))


def _days_since(iso: str | None) -> int | None:
    if not iso:
        return None
    from datetime import datetime, timezone
    try:
        d = datetime.fromisoformat(iso[:10]).replace(tzinfo=timezone.utc)
        return (datetime.now(timezone.utc) - d).days
    except ValueError:
        return None


def _clip(s: str | None, n: int = 90) -> str:
    if not s:
        return ""
    return s if len(s) <= n else s[: n - 1] + "…"


def build_report(rows, files_scanned, parse_errors, stale) -> str:
    def count(v):
        return sum(1 for r in rows if r["verdict"] == v)

    changed = [r for r in rows if r["verdict"] == "changed"]
    pending = [r for r in rows if r["verdict"] == "pending"]
    unreachable = [r for r in rows if r["verdict"] == "unreachable"]
    when = now_iso()[:16].replace("T", " ")
    L = [
        "# \U0001F33E Omnia data harvester",
        "",
        f"_Run {when} UTC · {files_scanned} data files · {len(rows)} cited sources_",
        "",
        "| | count |", "|---|--:|",
        f"| ⚠️ changed (confirmed) | {count('changed')} |",
        f"| \U0001F195 new | {count('new')} |",
        f"| ✅ unchanged | {count('unchanged')} |",
        f"| \U0001F440 pending (candidate, watching for a 2nd sighting) | {count('pending')} |",
        f"| \U0001F50C unreachable | {count('unreachable')} |",
        f"| \U0001F4C4 JSON parse errors | {len(parse_errors)} |",
        f"| \U0001F570️ stale (>{STALE_DAYS}d) | {len(stale)} |",
        "",
    ]
    if changed:
        L += ["## ⚠️ Sources CONFIRMED changed since last run", "",
              "Content differs from the baseline and held steady across two consecutive runs — review whether our data needs updating.", ""]
        for r in changed:
            L += [f"- **{_clip(r['title']) or r['url']}**", f"  - {r['url']}", f"  - cited by: {', '.join(r['citedBy'])}"]
        L.append("")
    if parse_errors:
        L += ["## \U0001F4C4 JSON parse errors (fix these — the data file is broken)", ""] + [f"- {p}" for p in parse_errors] + [""]
    if pending:
        L += ["<details><summary>\U0001F440 Pending — new content seen once; reported as changed only if it repeats next run (this is how per-request churn is filtered out)</summary>", ""]
        L += [f"- {_clip(r['title']) or r['url']} ({r['url']})" for r in pending]
        L += ["", "</details>", ""]
    if unreachable:
        L += ["<details><summary>\U0001F50C Unreachable this run (not a failure — retried next run)</summary>", ""]
        L += [f"- {_clip(r['title']) or r['url']} — {r['signal'].error or 'unreachable'}" for r in unreachable]
        L += ["", "</details>", ""]
    if stale:
        L += [f"<details><summary>\U0001F570️ Data files whose freshness marker is older than {STALE_DAYS} days</summary>", ""] + [f"- {s}" for s in stale] + ["", "</details>", ""]
    L += ["---", "_A CHANGED source is confirmed across two consecutive runs, so per-request churn (counters, timestamps) never counts; a real edit is caught on the next run. The harvester never fails on unreachable sources or changes._"]
    return "\n".join(L)


def main() -> None:
    scan = scan_repo(DATA_DIR)
    parse_errors = [f"{f.path}: {f.parse_error}" for f in scan.files if not f.ok]
    stale = [f"{f.path} (asOf {f.as_of}, {_days_since(f.as_of)}d ago)"
             for f in scan.files if f.ok and f.as_of and (_days_since(f.as_of) or 0) > STALE_DAYS]

    urls = [s.url for s in scan.sources]
    if NO_NETWORK:
        from probe import Signal
        signals = [Signal(url=u, ok=False, error="network skipped (--no-network)") for u in urls]
    else:
        signals = probe_all(urls, concurrency=10, timeout=15.0, retries=1)

    baseline = load_baseline(BASELINE_PATH)
    now = now_iso()
    rows = []
    for src, sig in zip(scan.sources, signals):
        verdict = "unreachable" if NO_NETWORK else classify(baseline.get(src.url), sig)
        rows.append({"verdict": verdict, "url": src.url, "signal": sig,
                     "citedBy": src.cited_by, "title": src.title})
        if not NO_NETWORK:
            baseline[src.url] = fold(baseline.get(src.url), sig, now, verdict)

    report = build_report(rows, len(scan.files), parse_errors, stale)
    with open(REPORT_PATH, "w", encoding="utf-8") as fh:
        fh.write(report + "\n")
    summary = os.environ.get("GITHUB_STEP_SUMMARY")
    if summary:
        try:
            with open(summary, "a", encoding="utf-8") as fh:
                fh.write(report + "\n")
        except OSError:
            pass
    if not NO_NETWORK:
        save_baseline(BASELINE_PATH, baseline)

    def c(v):
        return sum(1 for r in rows if r["verdict"] == v)
    print(f"\n\U0001F33E Harvester: {len(scan.files)} files, {len(rows)} sources — "
          f"⚠️ {c('changed')} changed · \U0001F195 {c('new')} new · "
          f"✅ {c('unchanged')} unchanged · \U0001F440 {c('pending')} pending · "
          f"\U0001F50C {c('unreachable')} unreachable · \U0001F4C4 {len(parse_errors)} parse errors · "
          f"\U0001F570️ {len(stale)} stale")
    print(f"   report → {REPORT_PATH}")

    if STRICT and c("changed") > 0:
        sys.exit(1)


if __name__ == "__main__":
    try:
        main()
    except Exception as e:  # never kill the workflow on an unexpected error unless --strict
        print(f"harvester: unexpected error: {e}", file=sys.stderr)
        if STRICT:
            sys.exit(1)
